validator/client/beacon-api/activation.go

```diff
@@ -0,0 +1,184 @@
+
+		for _, publicKey := range x.ValidatorActivationRequest.PublicKeys {
+			stringPubKey := hexutil.Encode(publicKey)
+			targetPubKeys[stringPubKey] = publicKey
+		}
+
+		stateValidators, err := x.getStateValidators(targetPubKeys)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get validators data")
+		}
+
+		for _, data := range stateValidators.Data {
+			pubkey, err := hexutil.Decode(data.Validator.PublicKey)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to read validator public key")
+			}
+
+			retrievedPubKeys[data.Validator.PublicKey] = struct{}{}
+
+			index, err := strconv.ParseUint(data.Index, 10, 64)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to read validator index")
+			}
+
+			validatorStatus, ok := beaconAPITogRPCValidatorStatus[data.Status]
+			if !ok {
+				return nil, errors.New("invalid validator status: " + data.Status)
+			}
+
+			statuses = append(statuses, &ethpb.ValidatorActivationResponse_Status{
+				PublicKey: pubkey,
+				Index:     types.ValidatorIndex(index),
+				Status:    &ethpb.ValidatorStatusResponse{Status: validatorStatus},
+			})
+		}
+
+		for targetStringPubKey, targetPubKey := range targetPubKeys {
+			if _, ok := retrievedPubKeys[targetStringPubKey]; !ok {
+				missingPubKeys[targetStringPubKey] = targetPubKey
+			}
+		}
+
+		for _, missingPubKey := range missingPubKeys {
+			statuses = append(statuses, &ethpb.ValidatorActivationResponse_Status{
+				PublicKey: missingPubKey,
+				Index:     types.ValidatorIndex(^uint64(0)),
+				Status:    &ethpb.ValidatorStatusResponse{Status: ethpb.ValidatorStatus_UNKNOWN_STATUS},
+			})
+		}
+
+		return &ethpb.ValidatorActivationResponse{
+			Statuses: statuses,
+		}, nil
+	case <-x.ctx.Done():
+		return nil, errors.New("context canceled")
+	}
+}
+
+func (x *beaconNodeValidatorWaitForActivationClient) getStateValidators(pubkeys map[string][]byte) (*rcpmiddleware.StateValidatorsResponseJson, error) {
+	params := neturl.Values{}
+
+	for pubkey := range pubkeys {
+		params.Add("id", pubkey)
+	}
+
+	url := buildURL(
+		x.beaconApiValidatorClient.url,
+		"eth/v1/beacon/states/head/validators",
+		params,
+	)
+
+	resp, err := x.beaconApiValidatorClient.httpClient.Get(url)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to query REST API validators endpoint")
+	}
+	defer func() {
+		if err = resp.Body.Close(); err != nil {
+			return
+		}
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		errorJson := &apimiddleware.DefaultErrorJson{}
+		if err := json.NewDecoder(resp.Body).Decode(&errorJson); err != nil {
+			return nil, errors.Wrap(err, "failed to decode response body validators error json")
+		}
+
+		return nil, errors.Errorf("error %d: %s", errorJson.Code, errorJson.Message)
+	}
+
+	validatorsJson := &rcpmiddleware.StateValidatorsResponseJson{}
+	if err := json.NewDecoder(resp.Body).Decode(&validatorsJson); err != nil {
+		return nil, errors.Wrap(err, "failed to decode response body validators json")
```

Comment:
```suggestion
		return nil, errors.Wrap(err, "failed to decode response json")
```
I think this is too cryptic

Reply:
Code deleted